### PR TITLE
TRITON-1991 factor finding an image out of AddServiceProcedure into steps.images.findSvcImg

### PR DESCRIPTION
**Pull request by trentm**

### CODE REVIEW

File: lib/steps/images.js [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+ * - `args.svcImg`: The found image manifest.
+ * - `args.needToDownloadImg`: Set to true if the image is not already in the DC
+ *   and needs to be downloaded from updates.joyent.com.
+ * - `args.channel`: Maybe. If updates.joyent.com was queried (i.e. the image
+ *   was not found locally), then this will be set to `args.channelArg` or
+ *   resolved to the default channel.
+ *
+ * If no matching image can be found, this callback with an error.
+ */
+function findSvcImg(args, cb) {
```
Comment (kusor) [inline lib/steps/images.js]: I know it's pretty obvious, but shouldn't the function check for `args` being an object and `cb` being a function first. Specially on the `args` case where you could find exceptions of the type `cannot get property args.svcName of undefined`

File: lib/steps/images.js [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+                    next();
+                    return;
+                }
+
+                sdcadm.imgapi.listImages({
+                    name: imgName
+                }, function (err, imgs) {
+                    if (err) {
+                        next(err);
+                    } else if (imgs && imgs.length > 0) {
```
Comment (kusor) [inline lib/steps/images.js]: Wouldn't be fine with `images.length` on the conditional?.

Also, I personally prefer to return as soon as possible and avoid multiple conditional branches. For example, on this case:

```
if (err) {
    next(err)
    return;
}
if (!imgs || !imgs.length) {
    next(new errors.UpdateError(format(
        'no "%s" image found in this DC\'s IMGAPI',
        imgName)));
    return;
}
// Assumption: these are already sorted by published_at.
ctx.svcImg = imgs[imgs.length - 1];
ctx.needToDownloadImg = false;
next();

```

As commented, this is just a personal preference but, for cases where the intermediate branch becomes more complex, I find it more clear this way. The next function `finding_Latest` is also an example of the same thing. Note you do the `if(imgs  && imgs.length)` check there.

File: lib/steps/images.js [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+                // Look first in the local DC IMGAPI.
+                sdcadm.imgapi.listImages({
+                    name: imgName,
+                    version: args.imgArg
+                }, function (localErr, localImgs) {
+                    if (localErr && !(localErr.body &&
+                        localErr.body.code === 'ResourceNotFound')) {
+                        next(localErr);
+                    } else if (!localErr && localImgs && localImgs.length > 0) {
```
Comment (kusor) [inline lib/steps/images.js]: Ditto. Within a conditional if it's true is b/c the length isn't zero.

File: lib/steps/images.js [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+                    } else {
+                        // Fallback to looking in updates.joyent.com.
+                        sdcadm.updates.listImages({
+                            name: imgName,
+                            version: args.imgArg,
+                            channel: ctx.channel
+                        }, function (updatesErr, updatesImgs) {
+                            if (updatesErr) {
+                                next(updatesErr);
+                            } else if (updatesImgs && updatesImgs.length > 0) {
```
Comment (kusor) [inline lib/steps/images.js]: Same thing :-)

File: lib/steps/images.js (hunk 2 above)
Comment (trentm) [reply lib/steps/images.js]: > Wouldn't be fine with images.length on the conditional?.

Yes. Either is fine.  Some languages (I think Rust is an example) require that you do a numeric comparator for this.  So I am letting some Rust-learning show through.  I can change if you feel strongly?

> Also, I personally prefer to return as soon as possible and avoid multiple conditional branches. ...


I agree if the non-error case is at all long. For better or worse, I tend to balance that with a desire to have one and only one "return" in a function: at the end.  That leads to my style sometimes being:

```
 if (any error case) {
     next(err);
    return;
}
// ... handle the non-error case
```

and sometimes:

```
if (error case 1) {
      // small error handling, call next(...)
} else if (error case 2) {
     // small error handling, call next(...)
} else {
    // *small* handling the non-error case, call next(...)
}
```


I can change if you feel strongly. Note that I was largely translating what was already in lib/procedures/add-service.js to refactor this, and I believe it was using the latter style.

File: lib/steps/images.js (hunk 1 above)
Comment (trentm) [reply lib/steps/images.js]: Sure, I can add that.

**Pull request merged**